spatial/r3/mat.go

```diff
@@ -0,0 +1,185 @@
+package r3
+
+import (
+	"gonum.org/v1/gonum/mat"
+)
+
+const (
+	badDim = "bad matrix dimensions"
+	badIdx = "bad matrix index"
+)
+
+// Mat represents a 3×3 matrix. Useful for rotation matrices and such.
+type Mat struct {
+	data [3][3]float64
+}
+
+var (
+	_ mat.Matrix = (*Mat)(nil)
+)
+
+// NewMat returns a new 3×3 matrix Mat type and populates it's elements
+// with values passed as argument in row-major form. If val argument
+// is nil then NewMat returns a matrix filled with zeros.
+func NewMat(val []float64) *Mat {
+	if len(val) != 9 {
+		if val == nil {
+			return &Mat{}
+		}
+		panic(badDim)
+	}
+	M := Mat{}
+	for iv := 0; iv < 9; iv++ {
+		i := iv / 3
+		j := iv % 3
+		M.data[i][j] = val[iv]
+	}
+	return &M
```

Comment:
Yeah I'm not fond of including unsafe in the main implementation. I know some Go based compilers break with it such as gopherjs. I'll see if I can split it up with build tags.

Reply:
It's worth noting that gopherjs is not a target. It would be good to keep it, but if that blocks other things we are OK with losing it. The raw matrix code is only possible with unsafe if we use a 3×3 array. If it's important to have unsafe-free code, then perhaps we consider using a `[9]float64`. That decision needs benchmarking.